### ISSUE TITLE
Better "stack too deep" errors

### ISSUE THREAD

**Issue by 134dd3v** [ethereum/solidity#12449]
## Abstract

```
Compiling 106 files with 0.8.11
CompilerError: Stack too deep when compiling inline assembly: Variable dataEnd is 1 slot(s) too deep inside the stack.
```

CompilerError does not mention the location where the problem is (or at least any hints where the problem could be).

## Motivation

I have a huge project, removed a parameter from a bunch of functions now the project doesn't compile and not sure where to look.

## Specification

If it tells which file / function causes this issue, or any hints it would be helpful.

## Backwards Compatibility

None

**Comment by chriseth** [comment]
This is an error inside a compiler-generated utility function. The problem is that the function can be generated from multiple places and it might even be that the error won't go away after you changed all the places it was generated from.

Did you activate the optimizer?

**Comment by 134dd3v** [comment]
Yes the optimiser is on. Sorry about the delay in response, that time, got the error fixed. I tried reverting the changes one by one and one of them made the the code compile. I think even if compiler shows list of "possible" places that could cause the issue, that still should help the dev.

**Comment by hrkrshnn** [comment]
Why don't we use the debug data part in Yul to print the location?

**Comment by chriseth** [comment]
Because it will print `#generated:4:5` or something - it won't be useful.

**Comment by cameel** [comment]
Maybe we could at least say which contract is being assembled?
In an example like this with 106 input files that would still be *something*.

**Comment by zerosnacks** [comment]
In my opinion this should not be tagged as low impact as in certain circumstances the stack to deep error does not resolve to a line or function in the source code which makes it almost impossible for users to track down where the stack to deep error originates from, especially when you rely on external dependencies.

Example: https://github.com/foundry-rs/foundry/issues/9674#issuecomment-2589712413

**Comment by cameel** [comment]
I agree. I think we should treat this with more urgency. It's been some time since the discussion above happened and it surfaced that there are actually a few things we could do to improve user experience here:
- At the very least the error should be caught and reported as a compilation error, not a failed assertion.
    - Reporting errors that happen during code generation used to be problematic but we recently needed it for EOF and we could easily switch to the same mechanism here: #15624.
- When compiling Solidity we can use the location of the contract as a fallback. In a big codebase that's still much better than nothing. In fact we recently started doing just that for unimplemented feature errors: #15643.
- I think we can still do better than that either via debug info or by tracking which function was last compiled or whatever. There are some ideas listed in #15139.

**Comment by cameel** [comment]
Linking #15740, which contains an example case where StackTooDeep is reported as an ICE rather than a proper error. Note that in some cases the compiler can and does report location - it's one of them.

**Comment by julianwells3510** [comment]
Yes, there’s some kind of givеаway going on. I got some Ethereum there yesterday for different wallets. It’s unlikely they’ll take it back from me, but it’s a pity I didn’t have more wallets and couldn’t get more. But maybe it’s for the best.

<img width="373" alt="Image" src="https://github.com/user-attachments/assets/6ff8db23-897c-42e7-a5f5-0ff32e4ee072" />

<img width="369" alt="Image" src="https://github.com/user-attachments/assets/0cc44043-3241-4743-9f1b-409b139c1023" />
